sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryHashPartitionVerifySuite.scala

```diff
@@ -0,0 +1,219 @@
+    }
+  }
+
+  private def getRandomSchemas(rand: Random): Seq[StructType] = {
+    (1 to numSchemaTypes).map { _ =>
+      RandomDataGenerator.randomNestedSchema(rand, maxNumFields, variableTypes)
+    }
+  }
+
+  private def getPartitionId(rows: Seq[UnsafeRow], hash: HashPartitioning): Seq[Int] = {
+    val partIdExpr = hash.partitionIdExpression
+    rows.map { row =>
+      partIdExpr.eval(row).asInstanceOf[Int]
+    }
+  }
+
+  test("SPARK-47788: Ensure the same hash function is used across batches.") {
+    val rowAndPartIdFile = new File(goldenFileURI.getPath, rowAndPartIdFilename)
+
+    if (regenerateGoldenFiles) {
+      // To limit the golden file size under 10Mb, please set the final val MAX_STR_LEN: Int = 100
+      // and final val MAX_ARR_SIZE: Int = 4 in org.apache.spark.sql.RandomDataGenerator
+
+      val seed = Random.nextInt()
+      val random = new Random(seed)
+      logInfo(s"Get random inputs with seed $seed")
+
+      val schemas = getRandomSchemas(random)
+
+      val os = new DataOutputStream(
+        codec.compressedOutputStream(new FileOutputStream(rowAndPartIdFile))
+      )
+
+      saveSchemas(schemas)
+
+      schemas.foreach { schema =>
+        // Streaming stateful ops rely on this distribution to partition the data.
+        // Spark should make sure this class's partition dependency remain unchanged.
+        val hash = StatefulOpClusteredDistribution(
+          Seq(BoundReference(0, schema, nullable = true)),
+          numShufflePartitions
+        ).createPartitioning(numShufflePartitions)
+
+        assert(sql/catalyst/src/test/scala/org/apache/spark/sql/RandomDataGenerator.scala
+          hash.isInstanceOf[HashPartitioning],
+          "StatefulOpClusteredDistribution should " +
+          "rely on HashPartitioning to ensure partitions remain the same for streaming " +
+          "stateful operators."
+        )
+        val rows = getRandomRows(schema, numRows, random)
+        val partitions = getPartitionId(rows, hash.asInstanceOf[HashPartitioning])
+        saveRowsAndPartIds(rows, partitions, os)
+      }
+      os.close()
+    } else {
+      val schemas = readSchemas()
+      val is = new DataInputStream(
+        codec.compressedInputStream(new FileInputStream(rowAndPartIdFile))
+      )
+      schemas.foreach { schema =>
+        val hash = StatefulOpClusteredDistribution(
+          Seq(BoundReference(0, schema, nullable = true)),
+          numShufflePartitions
+        ).createPartitioning(numShufflePartitions)
+
+        assert(
+          hash.isInstanceOf[HashPartitioning],
+          "StatefulOpClusteredDistribution should " +
+          "rely on HashPartitioning to ensure partitions remain the same for streaming " +
+          "stateful operators."
+        )
+
+        val (rows, expectedPartitions) = readRowsAndPartIds(is)
+        val partitions = getPartitionId(rows, hash.asInstanceOf[HashPartitioning])
+        assert(
+          partitions === expectedPartitions,
```

Comment:
Please consider what will be logged when this assertion fails, and whether the information is sufficient to investigate the issue. If there are only a few rows being shown as mismatched partitions, what would be the better format of assertion error message to understand the issue?

Reply:
Expanded the assertions to include number of total mismatch rows and samples